src/test/java/echoserver/EchoServerServiceTest.java

```diff
@@ -0,0 +1,130 @@
+package echoserver;
+
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.ArrayList;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+
+public class EchoServerServiceTest {
+
+    private ArrayList<FakeClientConnection> clients = new ArrayList<FakeClientConnection>();
+    private EchoServerService testServer;
+    private ServerSocket hostSocket;
+    private FakeHost host;
+    private Logger echoLogger;
+
+    class FakeHost implements ConnectionHub {
+        ServerSocket hostSocket;
+
+        FakeHost(ServerSocket hostSocket) {
+            this.hostSocket = hostSocket;
+        }
+
+        public ConnectionDataStream listenForClientConnection()  {
+            return clients.remove(0);
+        }
+
+        public void close() throws IOException {
+            hostSocket.close();
+        }
+
+        public Boolean isClosed() {
+            return false;
+        }
+    }
+
+    class FakeClientConnection implements ConnectionDataStream {
+        FakeClientConnection(Socket socket) {
+            this.socket = socket;
+        }
+
+        public void write(String message) {
+        }
+
+        public void close() {
+            try {
+                this.socket.close();
+            } catch (IOException ignore) {
+            }
+        }
+
+        public String readInput() {
+            return "";
+        }
+
+        public boolean detectEOF() {
+            return true;
+        }
+
+        public Socket socket;
+    }
+
+    @Before
+    public void setupService() throws IOException {
+        hostSocket = new ServerSocket(5000);
+        host = new FakeHost(hostSocket);
+        echoLogger = new Logger(new Console(new ByteArrayOutputStream()));
+    }
+
+    @Test
+    public void whenTheServerIsRunning_multipleClientsCanConnectAtOnce() throws IOException {
+        ThreadPoolExecutor threadHandler =
+                new ThreadPoolExecutor(100, 100, 100, TimeUnit.MILLISECONDS, new ArrayBlockingQueue<Runnable>(100));
+        testServer = new EchoServerService(host, echoLogger, threadHandler);
+        Socket clientASocket = new Socket();
+        Socket clientBSocket = new Socket();
+        FakeClientConnection clientA = new FakeClientConnection(clientASocket);
+        FakeClientConnection clientB = new FakeClientConnection(clientBSocket);
+
+        clientASocket.connect(hostSocket.getLocalSocketAddress());
+        clientBSocket.connect(hostSocket.getLocalSocketAddress());
+
+        clients.add(clientA);
+        clients.add(clientB);
+        clients.add(null);
```

Comment:
It took me a minute to figure out how `clients` is used. How could we make this simpler?

One approach I prefer over mutating state is to construct my fake with all the data it needs from the get-go

```java
class FakeHost implements ConnectionHub {
        List<ConnectionDataStream> connections;
        boolean isClosed = false;

        FakeHost(List<ConnectionDataStream> connections) {
            this.connections = connections;
        }

        public ConnectionDataStream listenForClientConnection()  {
            return connections.remove(0);
        }

        public void close() throws IOException {
            isClosed = true;
        }

        public Boolean isClosed() {
            return isClosed;
        }
    }
```

This lets allows the following

```java
new FakeHost(Array.asList(new FakeConnection(), new FakeConnection(), null));
```